mediagrains/gsf.py

```diff
@@ -151,310 +152,497 @@ def __init__(self, msg, i, major, minor):
-        """Decode a GSF formatted bytes object, returning a dictionary mapping
-        sequence ids to lists of GRAIN objects (or subclasses of such)."""
-        b = bytes(s)
-        i = 0
+    def decode_file_headers(self):
+        """Verify the file is a supported version, and get the file header
 
-        (major, minor, i) = self._decode_ssb_header(b, i)
+        :returns: File header data (segments and tags) as a dict
+        :raises GSFDecodeBadVersionError: If the file version is not supported
+        :raises GSFDecodeBadFileTypeError: If this isn't a GSF file
+        :raises GSFDecodeError: If the file doesn't have a "head" block
+        """
+        (major, minor) = self._decode_ssb_header()
         if (major, minor) != (7, 0):
             raise GSFDecodeBadVersionError("Unknown Version {}.{}".format(major, minor), 0, major, minor)
 
-        (head, i) = self._decode_head(b, i)
-        segments = {}
+        try:
+            with GSFBlock(self.file_data, want_tag="head") as head_block:
+                return self._decode_head(head_block)
+        except EOFError:
+            raise GSFDecodeError("No head block found in file", self.file_data.tell())
 
-        while i < len(b):
-            (grain, local_id, i) = self._decode_grai(b, i)
+    def grains(self, skip_data=False):
+        """Generator to get grains from the GSF file. Skips blocks which aren't "grai".
 
-            if grain is None:
-                break
+        :param skip_data: If True, grain data blocks will be seeked over and only grain headers will be read
+        :yields: (Grain, local_id) tuple for each grain
+        :raises GSFDecodeError: If grain is invalid (e.g. no "gbhd" child)
+        """
+        while True:
+            try:
+                with GSFBlock(self.file_data, want_tag="grai") as grai_block:
+                    if grai_block.size == 0:
+                        return  # Terminator block reached
+
+                    local_id = grai_block.read_uint(2)
+                    with GSFBlock(self.file_data, want_tag="gbhd", raise_on_wrong_tag=True) as gbhd_block:
+                        meta = self._decode_gbhd(gbhd_block)
+
+                    data = None
+
+                    if grai_block.has_child_block() and not skip_data:
+                        with GSFBlock(self.file_data, want_tag="grdt") as grdt_block:
+                            if grdt_block.get_remaining() > 0:
+                                data = self.file_data.read(grdt_block.get_remaining())
+
+                    yield (self.Grain(meta, data), local_id)
+            except EOFError:
+                return  # We ran out of grains to read and hit EOF
+
+    def decode(self, s):
+        """Decode a GSF formatted bytes object
+
+        :param s: GSF-formatted bytes object
+        :returns: A dictionary mapping sequence ids to lists of GRAIN objects (or subclasses of such).
+        """
+        if (s):
+            self.file_data = BytesIO(s)
+
```

Comment:
This is slightly odd, it seems that you can call decode(False) to use an already set value for self.file_data ... did you mean to give s a default value of None or something?

Reply:
Drat

Reply:
Fixed in dd6af1b